File: bindings/python/vietasr/result.py
# Keep annotations lazy so `dict[...]` / `list[...]` generics do not need
# evaluation at runtime — required to import on Python 3.8.
from __future__ import annotations

import json
from typing import Any


class Result:
    def __init__(self, payload: dict[str, Any]):
        self.payload = payload

    @classmethod
    def from_json(cls, raw: str) -> "Result":
        return cls(json.loads(raw or "{}"))

    @property
    def text(self) -> str:
        return self.payload.get("text", "")

    @property
    def partial(self) -> str:
        return self.payload.get("partial", "")

    @property
    def is_final(self) -> bool:
        return bool(self.payload.get("is_final", False))

    @property
    def segments(self) -> list[dict[str, Any]]:
        return self.payload.get("segments", [])

    @property
    def speakers(self) -> list[dict[str, Any]]:
        return self.payload.get("speakers", [])

    def field(self, key: str) -> Any:
        return self.payload.get(key)

    def to_json(self) -> str:
        return json.dumps(self.payload, ensure_ascii=False)

    def __getitem__(self, key: str) -> Any:
        return self.payload[key]

    def __contains__(self, key: str) -> bool:
        return key in self.payload

    def __repr__(self) -> str:
        return f"Result({self.payload!r})"

File: bindings/csharp/Vietasr/Result.cs
using System.Text.Json;

namespace Vietasr
{
    public sealed class Result
    {
        public string Text { get; }
        public string Partial { get; }
        public bool IsFinal { get; }
        public string Raw { get; }

        private readonly JsonElement root;

        private Result(string raw, JsonElement root)
        {
            Raw = raw;
            this.root = root;
            Text = GetString("text");
            Partial = GetString("partial");
            IsFinal = root.ValueKind == JsonValueKind.Object
                      && root.TryGetProperty("is_final", out var f)
                      && f.ValueKind == JsonValueKind.True;
        }

        internal static Result FromJson(string raw)
        {
            if (string.IsNullOrEmpty(raw))
            {
                using var empty = JsonDocument.Parse("{}");
                return new Result("{}", empty.RootElement.Clone());
            }
            try
            {
                using var doc = JsonDocument.Parse(raw);
                return new Result(raw, doc.RootElement.Clone());
            }
            catch (JsonException)
            {
                using var empty = JsonDocument.Parse("{}");
                return new Result(raw, empty.RootElement.Clone());
            }
        }

        public JsonElement? Field(string key)
        {
            if (root.ValueKind == JsonValueKind.Object
                && root.TryGetProperty(key, out var value))
            {
                return value;
            }
            return null;
        }

        public string ToJson() => Raw;

        public override string ToString() => Text;

        private string GetString(string key)
        {
            if (root.ValueKind == JsonValueKind.Object
                && root.TryGetProperty(key, out var value)
                && value.ValueKind == JsonValueKind.String)
            {
                return value.GetString() ?? string.Empty;
            }
            return string.Empty;
        }
    }
}
